onmetal-controller-manager/controllers/networking/natgateway_controller_test.go

```diff
@@ -0,0 +1,278 @@
+			},
+			Spec: networkingv1alpha1.NATGatewaySpec{
+				Type: networkingv1alpha1.NATGatewayTypePublic,
+				IPFamilies: []corev1.IPFamily{
+					corev1.IPv4Protocol,
+				},
+				IPs: []networkingv1alpha1.NATGatewayIP{
+					{
+						Name: "ip1",
+					},
+				},
+				NetworkRef: corev1.LocalObjectReference{
+					Name: network.Name,
+				},
+				NetworkInterfaceSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"foo": "bar"},
+				},
+				PortsPerNetworkInterface: pointer.Int32(2),
+			},
+		}
+		Expect(k8sClient.Create(ctx, natGateway)).To(Succeed())
+
+		By("waiting for the nat gateway routing to exist with no destinations")
+		natGatewayKey := client.ObjectKeyFromObject(natGateway)
+		natGatewayRouting := &networkingv1alpha1.NATGatewayRouting{}
+		Eventually(func(g Gomega) {
+			err := k8sClient.Get(ctx, natGatewayKey, natGatewayRouting)
+			Expect(client.IgnoreNotFound(err)).NotTo(HaveOccurred())
+			g.Expect(err).NotTo(HaveOccurred())
+
+			g.Expect(metav1.IsControlledBy(natGatewayRouting, natGateway)).To(BeTrue(), "nat gateway routing is not controlled by nat gateway: %#v", natGatewayRouting.OwnerReferences)
+			g.Expect(natGatewayRouting.Destinations).To(BeEmpty())
+		}).Should(Succeed())
+
+		natGatewayBase := natGateway.DeepCopy()
+		ip := netip.IPv4Unspecified()
+		for _, v := range natGateway.Spec.IPs {
+			natGateway.Status.IPs = append(natGateway.Status.IPs, networkingv1alpha1.NATGatewayIPStatus{
+				Name: v.Name,
+				IP: commonv1alpha1.IP{
+					Addr: ip.Next(),
+				},
+			})
+		}
+		Expect(k8sClient.Patch(ctx, natGateway, client.MergeFrom(natGatewayBase))).To(Succeed())
+
+		By("creating a network interface")
+		nic := &networkingv1alpha1.NetworkInterface{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    ns.Name,
+				GenerateName: "nic-",
+				Labels:       map[string]string{"foo": "bar"},
+			},
+			Spec: networkingv1alpha1.NetworkInterfaceSpec{
+				NetworkRef: corev1.LocalObjectReference{Name: network.Name},
+				IPFamilies: []corev1.IPFamily{
+					corev1.IPv4Protocol,
+				},
+				IPs: []networkingv1alpha1.IPSource{
+					{
+						Value: commonv1alpha1.MustParseNewIP("10.0.0.1"),
+					},
+				},
+			},
+		}
+		Expect(k8sClient.Create(ctx, nic)).To(Succeed())
+
+		By("waiting for the nat gateway routing to be updated")
+		Eventually(func(g Gomega) {
+			Expect(k8sClient.Get(ctx, natGatewayKey, natGatewayRouting)).To(Succeed())
+
+			g.Expect(natGatewayRouting.NetworkRef.Name).To(BeEquivalentTo(network.Name))
+			g.Expect(natGatewayRouting.NetworkRef.UID).To(BeEquivalentTo(network.UID))
+
+			g.Expect(natGatewayRouting.Destinations).To(HaveLen(1))
+
+			g.Expect(natGatewayRouting.Destinations[0].Name).To(BeEquivalentTo(nic.Name))
+			g.Expect(natGatewayRouting.Destinations[0].UID).To(BeEquivalentTo(nic.UID))
+			g.Expect(natGatewayRouting.Destinations[0].IPs).To(HaveLen(1))
+			g.Expect(natGatewayRouting.Destinations[0].IPs[0].IP).To(BeEquivalentTo(natGateway.Status.IPs[0].IP))
```

Comment:
We should not use `BeEquivalentTo` unless really necessary. `BeEquivalentTo` lets mismatching types (e.g. from type aliases) slip through undiscovered. In general `Equal` is the assertion you want to make.